lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package storage
+
+import (
+	"bytes"
+	"sort"
+	"sync"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"github.com/ChainSafe/gossamer/pkg/trie"
+)
+
+// storageDiff is a structure that stores the differences between consecutive
+// states of a trie, such as those occurring during the execution of a block.
+// It records updates (upserts), deletions, and changes to child tries.
+// This mechanism facilitates applying state transitions efficiently.
+// Changes accumulated in storageDiff can be applied to a trie using
+// the `applyToTrie` method
+type storageDiff struct {
+	upserts        map[string][]byte
+	deletes        map[string]bool
+	childChangeSet map[string]*storageDiff
+	mtx            sync.RWMutex
+}
+
+// newChangeSet initialises and returns a new storageDiff instance
+func newStorageDiff() *storageDiff {
+	return &storageDiff{
+		upserts:        make(map[string][]byte),
+		deletes:        make(map[string]bool),
+		childChangeSet: make(map[string]*storageDiff),
+		mtx:            sync.RWMutex{},
+	}
+}
+
+// get retrieves the value associated with the key if it's present in the
+// change set and returns a boolean indicating if the key is marked for deletion
+func (cs *storageDiff) get(key string) ([]byte, bool) {
+	if cs == nil {
+		return nil, false
+	}
+
+	cs.mtx.RLock()
+	defer cs.mtx.RUnlock()
+
+	// Check in recent upserts if not found check if we want to delete it
+	if val, ok := cs.upserts[key]; ok {
+		return val, false
+	} else if deleted := cs.deletes[key]; deleted {
+		return nil, true
+	}
+
+	return nil, false
+}
+
+// upsert records a new value for the key, or updates an existing value.
+// If the key was previously marked for deletion, that deletion is undone
+func (cs *storageDiff) upsert(key string, value []byte) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	// If we previously deleted this trie we have to undo that deletion
+	if cs.deletes[key] {
+		delete(cs.deletes, key)
+	}
```

Comment:
check if the key is present, or delete the key directly `delete(map, key)`
```suggestion
	// If we previously deleted this trie we have to undo that deletion
	if _, ok := cs.deletes[key]; ok {
		delete(cs.deletes, key)
	}
```

Reply:
Based on my previous [comment](https://github.com/ChainSafe/gossamer/pull/3777#discussion_r1534424586) I think is more intuitive to keep it with `map[string]bool` in this case the `if` will return `false` if the key does not exists